sqlserver/datadog_checks/sqlserver/schemas.py

```diff
@@ -0,0 +1,369 @@
+try:
+    import datadog_agent
+except ImportError:
+    from ..stubs import datadog_agent
+
+import copy
+import json
+import time
+
+from datadog_checks.base.utils.db.utils import default_json_event_encoding
+from datadog_checks.base.utils.tracking import tracked_method
+from datadog_checks.sqlserver.const import (
+    COLUMN_QUERY,
+    DB_QUERY,
+    FOREIGN_KEY_QUERY,
+    INDEX_QUERY,
+    PARTITIONS_QUERY,
+    SCHEMA_QUERY,
+    STATIC_INFO_ENGINE_EDITION,
+    STATIC_INFO_VERSION,
+    TABLES_IN_SCHEMA_QUERY,
+)
+from datadog_checks.sqlserver.utils import execute_query_output_result_as_a_dict, get_list_chunks
+
+
+class SubmitData:
+    MAX_COLUMN_COUNT = 10_000
+
+    # TBD - REDAPL has a 3MB limit per resource
+    # If a column payload is ~ 10bytes : name, type, default , if nullable nullable
+    # then the limit should be only 25_000.
+
+    MAX_TOTAL_COLUMN_COUNT = 100_000
+
+    def __init__(self, submit_data_function, base_event, logger):
+        self._submit_to_agent_queue = submit_data_function
+        self._base_event = base_event
+        self._log = logger
+
+        self._columns_count = 0
+        self._total_columns_count = 0
+        self.db_to_schemas = {}  # dbname : { id : schema }
+        self.db_info = {}  # name to info
+
+    def set_base_event_data(self, hostname, tags, cloud_metadata, dbms_version):
+        self._base_event["host"] = hostname
+        self._base_event["tags"] = tags
+        self._base_event["cloud_metadata"] = cloud_metadata
+        self._base_event["dbms_version"] = dbms_version
+
+    def reset(self):
+        self._columns_count = 0
+        self._total_columns_count = 0
+        self.db_to_schemas = {}
+        self.db_info = {}
+
+    def store_db_info(self, db_name, db_info):
+        self.db_info[db_name] = db_info
+
+    def store(self, db_name, schema, tables, columns_count):
+        self._columns_count += columns_count
+        self._total_columns_count += columns_count
+        schemas = self.db_to_schemas.setdefault(db_name, {})
+        if schema["id"] in schemas:
+            known_tables = schemas[schema["id"]].setdefault("tables", [])
+            known_tables = known_tables + tables
+        else:
+            schemas[schema["id"]] = copy.deepcopy(schema)
+            schemas[schema["id"]]["tables"] = tables
+        if self._columns_count > self.MAX_COLUMN_COUNT:
+            self._submit()
```

Comment:
I'm having a lot of trouble understanding the flow of this code. I think its because `submit` can be called in a variety of circumstances. It might be helpful to add a comment describing those scenarios or (if possible) scoping all of the submit decisions to a single method? 

cc @lu-zhengda any ideas?